library/core/src/num/mod.rs

```diff
@@ -846,37 +870,67 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+    if radix <= 16 && src.len() <= mem::size_of::<T>() * if radix <= 8 { 1 } else { 2 } - if is_signed_ty { 1 } else { 0 } {
```

Comment:
```suggestion
    if radix <= 16 && src.len() <= mem::size_of::<T>() * if radix <= 8 { 1 } else { 2 } - if is_signed_ty { 1 } else { 0 } {
```
~~Then what about 2? It could be 4. Ideally we would want to use a match or conditionals for this. Like `<= 16 = 2`, `<= 8 = 4`, `= 2 = 16`~~

We can probably just ignore that branch and use 16 as a baseline, easier to just multiply by 2.

Reply:
I'd argue base `2` probably wouldn't work nicely anyhow. If your decoding binary then you're probably putting it in a type that would be exactly the right size and so this trick wouldn't work.

Maybe for types T < usize, we parse them into usize and then cast to T with one overflow check - that way for those types we could accept the full range of T unchecked. As it is I suspect most useful numbers will get a speedup, so not sure if that's that helpful.

Reply:
@joshtriplett I mentioned multiplying by 2 here, I thought it could reduce a branch while being able to cover a lot of the general cases, it could be faster. Maybe we could run a benchmark with two of them.